lib/Dialect/FIRRTL/Transforms/LowerLayers.cpp

```diff
@@ -58,13 +94,79 @@ FModuleOp LowerLayersPass::buildNewModule(OpBuilder &builder, Location location,
   return newModule;
 }
 
-/// Process a module to remove any layer blocks it has.
-void LowerLayersPass::runOnModule(FModuleOp moduleOp) {
+void LowerLayersPass::removeLayersFromValue(Value value) {
+  auto type = dyn_cast<RefType>(value.getType());
+  if (!type || !type.getLayer())
+    return;
+  value.setType(type.removeLayer());
+}
+
+void LowerLayersPass::removeLayersFromRefCast(RefCastOp cast) {
+  auto result = cast.getResult();
+  auto oldType = result.getType();
+  if (oldType.getLayer()) {
+    auto input = cast.getInput();
+    auto srcType = input.getType();
+    auto newType = oldType.removeLayer();
+    if (newType == srcType) {
+      result.replaceAllUsesWith(input);
+      cast->erase();
+      return;
+    }
+    result.setType(newType);
+  }
+}
+
+void LowerLayersPass::removeLayersFromPorts(FModuleLike moduleLike) {
+  auto oldTypeAttrs = moduleLike.getPortTypesAttr();
+  SmallVector<Attribute> newTypeAttrs;
+  newTypeAttrs.reserve(oldTypeAttrs.size());
+  bool changed = false;
+
+  for (auto typeAttr : oldTypeAttrs.getAsRange<TypeAttr>()) {
+    if (auto refType = dyn_cast<RefType>(typeAttr.getValue())) {
+      if (refType.getLayer()) {
+        typeAttr = TypeAttr::get(refType.removeLayer());
+        changed = true;
+      }
+    }
+    newTypeAttrs.push_back(typeAttr);
+  }
+
+  if (!changed)
+    return;
+
+  moduleLike->setAttr(FModuleLike::getPortTypesAttrName(),
+                      ArrayAttr::get(moduleLike.getContext(), newTypeAttrs));
+
+  if (auto moduleOp = dyn_cast<FModuleOp>(moduleLike.getOperation())) {
+    for (auto arg : moduleOp.getBodyBlock()->getArguments())
+      removeLayersFromValue(arg);
+  }
+}
+
+void LowerLayersPass::runOnModuleLike(FModuleLike moduleLike) {
   LLVM_DEBUG({
-    llvm::dbgs() << "Module: " << moduleOp.getModuleName() << "\n";
+    llvm::dbgs() << "Module: " << moduleLike.getModuleName() << "\n";
     llvm::dbgs() << "  Examining Layer Blocks:\n";
   });
 
+  // Strip away layers from the interface of the module-like op.
+  TypeSwitch<Operation *, void>(moduleLike.getOperation())
+      .Case<FModuleOp>([&](auto op) {
+        op.setLayers({});
+        removeLayersFromPorts(op);
+        runOnModuleBody(op);
+      })
+      .Case<FModuleOp, FExtModuleOp, FIntModuleOp, FMemModuleOp>([&](auto op) {
+        op.setLayers({});
+        removeLayersFromPorts(op);
+      })
```

Comment:
Remove stray `FModuleOp` double match:

```suggestion
      .Case<FExtModuleOp, FIntModuleOp, FMemModuleOp>([&](auto op) {
        op.setLayers({});
        removeLayersFromPorts(op);
      })
```

It would be nice if there was `TypeSwitch` with fall through to collapse these...